src/app/static/src/js/components/reports/reportReadersList.vue

```diff
@@ -2,29 +2,68 @@
     <div id="report-readers-list">
         <label class="font-weight-bold">Report readers</label>
         <div>
+            <div class="input-group mb-3">
+                <input v-model="add_email" class="form-control form-control-sm" type="text" placeholder="user email" value />
+                <div class="input-group-append">
+                    <button v-on:click="add" type="submit" class="btn btn-sm">Add reader</button>
+                </div>
+            </div>
             <ul class="list-unstyled report-readers">
-                <li v-for="reader in readers">
+                <li v-for="reader in readers" v-bind:id="reader.username">
                     <span>{{reader.display_name}}</span>
-                    <div class="text-muted small" style="margin-top: -1rem">{{reader.username}}</div>
+                    <div class="text-muted small" style="margin-top: -1rem">{{reader.email}}</div>
                 </li>
             </ul>
         </div>
+        <div class="text-danger mt-3" v-if="error.length > 0">
+            {{error}}
+        </div>
     </div>
 </template>
 
 <script>
+    import {api} from "../../utils/api";
 
     export default {
         name: 'reportReadersList',
-        props: ['report', 'readers'],
+        props: ['report', 'initial_readers'],
         data() {
             return {
-                error: ""
+                add_email: "",
+                error: "",
+                readers: []
             }
         },
+        mounted() {
+            this.refreshReaders();
+        },
         methods: {
             add: function () {
-                //TODO!
+                const data = {
+                    name: "reports.read",
+                    action: "add",
+                    scope_prefix: "report",
+                    scope_id: this.report.name
+                };
+
+                api.post(`/users/${encodeURIComponent(this.add_email)}/actions/associate-permission/`, data)
+                    .then(() => {
+                        this.refreshReaders();
+                        this.add_email = "";
+                        this.error = "";
+                    })
+                    .catch(() => {
+                        this.error = "Error: could not add reader";
```

Comment:
If this fails because the provided email is not valid we should get back something informative to that effect right? And in general if this call fails we should get a Result object back with an error message of some kind. Would be nice to return that to the UI here and just fallback to this generic error message if the response is malformed. Similarly below.